DeviceSimulation/SimulationTests.cs

```diff
@@ -1,33 +1,299 @@
+            var response = this.httpClient.PostAsync(request).Result;
+
+            //Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            var getCurrentSimulationRequest = new HttpRequest(Constants.DS_ADDRESS + "/simulations/1");
+            var getCurrentSimulationResponse = this.httpClient.GetAsync(getCurrentSimulationRequest).Result;
+            Assert.Equal(HttpStatusCode.OK, getCurrentSimulationResponse.StatusCode);
+        }
+
+        [Fact, Trait("Type", "IntegrationTest")]
+        public async void Should_Create_SimulatedDevice()
+        {
+            //Arrange
+            this.Should_Delete_Existing_Simulation();
+            var simulation = JObject.Parse(@"{   
+                'Enabled': true, 
+                'DeviceModels': [   
+                    {   
+                        'Id': 'truck-01', 
+                        'Count': 5 
+                    } 
+                ]
+            }");
+
+            //Act
+            var request = new HttpRequest(Constants.DS_ADDRESS + "/simulations");
+            request.SetContent(simulation);
+            var response = this.httpClient.PostAsync(request).Result;
+
+            //Assert
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+            Thread.Sleep(Constants.WAIT_TIME);
+            RegistryManager registry = RegistryManager.CreateFromConnectionString(IOTHUB_CONNECTION_STRING);
+            Twin deviceTwin = await registry.GetTwinAsync("truck-01.0");
+            Assert.True(deviceTwin != null && deviceTwin.Tags["IsSimulated"] == 'Y');
+        }
+
+        [Fact, Trait("Type", "IntegrationTest")]
+        public async void Should_Upgrade_Firmware_On_SimulatedDevice()
+        {
+            //Arrange
+            this.Should_Delete_Existing_Simulation();
+            var simulation = JObject.Parse(@"{   
+                'Enabled': true, 
+                'DeviceModels': [   
+                    {   
+                        'Id': 'truck-01', 
+                        'Count': 5 
+                    } 
+                ]
+            }");
+            var request = new HttpRequest(Constants.DS_ADDRESS + "/simulations");
+            request.SetContent(simulation);
+            var response = this.httpClient.PostAsync(request).Result;
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+
+            Thread.Sleep(Constants.WAIT_TIME);
+
+            //Act
+            ServiceClient serviceClient = ServiceClient.CreateFromConnectionString(IOTHUB_CONNECTION_STRING);
+            Task<CloudToDeviceMethodResult> firmwareUpgradeMethodResult = serviceClient.InvokeDeviceMethodAsync("truck-01.0",
+                new CloudToDeviceMethod("FirmwareUpdate"));
+            CloudToDeviceMethodResult firmwareUpdateResponse = await firmwareUpgradeMethodResult.ConfigureAwait(false);
+
+            //Assert
+            Assert.Equal(200, firmwareUpdateResponse.Status);
```

Comment:
This is only verifying that the method was called, not that it worked. The firmware-upgrade method takes several seconds to verify. The test should wait for the method to complete, then query the reported firmware version to verify that the method worked properly.

Reply:
+1